### PR TITLE
Hotseat password

### PR DESCRIPTION
**Pull request by alexban011**
https://user-images.githubusercontent.com/21990802/176562948-1f173737-56f2-4bdd-a650-051f8c83d5f0.mp4

### CODE REVIEW

**Comment by alexban011** [comment]
Clicking anywhere switches screeens, but that might not be obvious when you see a button on the screen. Should I add a button as well to avoid confusion?

**Comment by alexban011** [comment]
> password maybe is not nesesary for hotseat

Well its a civ V feature and I thought they probably had their reason for adding it.

File: core/src/com/unciv/ui/worldscreen/PlayerReadyScreen.kt [hunk 1]
```diff
@@ -4,24 +4,113 @@ import com.badlogic.gdx.scenes.scene2d.Touchable
...
+                table.add(savePasswordButton).padTop(10f).row()
+        }
+
+        val removePasswordButton = "Remove password".toTextButton()
+        removePasswordButton.onClick {
+            if (hash(enteredPassword) == curCiv.hotseatPassword) {
+                curCiv.hotseatPassword = ""
+            } else {
+                startTurn = false
+                ToastPopup("You need to enter the current password first!", this)
```
Comment (yairm210) [inline core/src/com/unciv/ui/worldscreen/PlayerReadyScreen.kt:54]: "Wrong password!" is used below, reuse is recommended

File: core/src/com/unciv/ui/worldscreen/PlayerReadyScreen.kt [hunk 2]
```diff
@@ -4,24 +4,113 @@ import com.badlogic.gdx.scenes.scene2d.Touchable
...
 import com.unciv.ui.utils.extensions.onClick
 import com.unciv.ui.utils.extensions.toLabel
-import com.unciv.utils.concurrency.Concurrency
+import com.unciv.ui.utils.extensions.toTextButton
+import java.security.MessageDigest
 
 class PlayerReadyScreen(worldScreen: WorldScreen) : BaseScreen() {
+    var startTurn = true
+    val curCiv = worldScreen.viewingCiv
+    var enteredPassword = curCiv.hotseatPassword
```
Comment (yairm210) [inline core/src/com/unciv/ui/worldscreen/PlayerReadyScreen.kt:18]: ???? WHAT????

File: core/src/com/unciv/ui/worldscreen/PlayerReadyScreen.kt [hunk 3]
```diff
@@ -4,24 +4,113 @@ import com.badlogic.gdx.scenes.scene2d.Touchable
...
 import com.unciv.ui.utils.BaseScreen
+import com.unciv.ui.utils.UncivTextField
 import com.unciv.ui.utils.extensions.onClick
 import com.unciv.ui.utils.extensions.toLabel
-import com.unciv.utils.concurrency.Concurrency
+import com.unciv.ui.utils.extensions.toTextButton
+import java.security.MessageDigest
 
 class PlayerReadyScreen(worldScreen: WorldScreen) : BaseScreen() {
+    var startTurn = true
```
Comment (yairm210) [inline core/src/com/unciv/ui/worldscreen/PlayerReadyScreen.kt:16]: What does this actually mean? Unclear

Comment (alexban011) [reply core/src/com/unciv/ui/worldscreen/PlayerReadyScreen.kt:16]: The entire screen has an onclick{} which changes the screen. I made this so I can disable that when clicking on the Textbox or on the buttons. Is there a cleaner way of doing this? I tried overriding the onClick but it still kept the one from the table.

File: core/src/com/unciv/ui/worldscreen/PlayerReadyScreen.kt (hunk 2 above)
Comment (alexban011) [reply core/src/com/unciv/ui/worldscreen/PlayerReadyScreen.kt]: I renamed these to be more clear, but I'm not sure if this resolves this conversation :)

**Comment by alexban011** [comment]
I've been thinking if I should even keep working on this. A discord poll would probably be useless since people will probably vote yes and never use the feature.

Adding an option when creating the game would make this look cleaner but I'm not sure I like the idea of adding yet another option which probably is not gonna be used. I could add the password field underneath the nations (like the user ID fields) and use something like "Leave empty for no password", but that probably adds even more clutter.

So I'd like to know if I should keep doing this, even though its a Civ 5 feature, I'm not sure if this would be an improvement or just clutter in the NewGameScreen

**Comment by MindaugasRumsa51** [comment]
i tested such feature with 2 hotseaters (hotlyers cause they lying and giving phone to each other)
i would say - it is very useful. also if no password set - confirm dialog needed for accident touches of screen

**Comment by alexban011** [comment]
Closing this as `too-lazy-to-continue` :D

**Comment by touhidurrr** [comment]
@ofetios @Caballero-Arepa, want to work on it again together with support for passwords in UncivServer.xyz? I have time.

**Comment by Caballero-Arepa** [comment]
What would I have to do? I don't know how to code😬

**Comment by touhidurrr** [comment]
> What would I have to do? I don't know how to code😬

Oh, I thought you were interested since you recently forked https://github.com/touhidurrr/UncivServer.xyz and asked about passwords support on discord.

**Comment by ofetios** [comment]
> @ofetios @Caballero-Arepa, want to work on it again together with support for passwords in **UncivServer.xyz**? I have time. The main problem is enforcing passwords serverside. Should be easier now since **JSON.parse()** works.

If you pinged me for permission to use anything that's in this PR then yeah go ahead feel free to use whatever code you want. :)

**Comment by touhidurrr** [comment]
> If you pinged me for permission to use anything that's in this PR then yeah go ahead feel free to use whatever code you want. :)

Not really, Previously i tried implementing passwords in UncivServer.xyz and it did not work out well. So, I thought we can collaborate to get passwords working again in a better way.
Basically I am interested in implementing 2 things for now:
1. Password protected games.
2. Password protected user ids.

**Comment by Caballero-Arepa** [comment]
> > What would I have to do? I don't know how to code😬
> 
> Oh, I thought you were interested since you recently forked https://github.com/touhidurrr/UncivServer.xyz and asked about passwords support on discord.

Oh yes, I'm interested, and i would like to help with what I can, which I don't know what is.

**Comment by ofetios** [comment]
> I thought we can collaborate to get passwords working again in a better way

I'm very busy rn with faculty and work so it'd prob take me some time to get up to date with the new changes. I'll try but don't depend on me

> Password protected games

Well this is kinda what I implemented here, just for hotseat only really, maybe needs some modifications to work on online games 

> Password protected user ids.

How would you put passwords on this though? Maybe whenever people click on the multiplayer button they get a popup to choose a password, and the user id is generated based on that password? Might not work though I kinda forgot how exactly this system works.

**Pull request closed without merging**